Track GET/LIST GoogleCloudStorage API calls

fcofdez opened the pull request:
Adds tracking for the API calls performed by the GoogleCloudStorage
underlying SDK. It hooks an HttpResponseInterceptor to the SDK
transport layer and does http request filtering based on the URI
paths that we are interested to track. Unfortunately we cannot hook
a wrapper into the ServiceRPC interface since we're using different
levels of abstraction to implement retries during reads
(GoogleCloudStorageRetryingInputStream).

fcofdez committed:
Track GET/LIST GoogleCloudStorage API calls

Adds tracking for the API calls performed by the GoogleCloudStorage
underlying SDK. It hooks an HttpResponseInterceptor to the SDK
transport layer and does http request filtering based on the URI
paths that we are interested to track. Unfortunately we cannot hook
a wrapper into the ServiceRPC interface since we're using different
levels of abstraction to implement retries during reads
(GoogleCloudStorageRetryingInputStream).

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

fcofdez commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -411,4 +411,14 @@ private static String buildKey(String keyPath, String s) {
+        try {
+            return storageService.stats(clientName).toMap();
+        } catch (IOException e) {
+            logger.error(() -> new ParameterizedMessage("unable to get GCS usage stats for client [{}]", clientName), e);
```
I'm not really sure if we should ignore the exception here or propagate it.

fcofdez commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -173,4 +198,107 @@ static Integer toTimeout(final TimeValue timeout) {
         return Math.toIntExact(timeout.getMillis());
     }
 
+    private static class StorageService {
```
I would appreciate some feedback on better naming here.

original-brownbear reviewed (commented):
Thanks @fcofdez the approach in general looks good (glad to see it working end-to-end with the interceptor:)).  Our way of handling the clients is interfering with the implementation a little.

Ideally, I think we should find a way of tying the lifecycle of the `GoogleCloudStorageOperationsStats` to that of `GoogleCloudStorageBlobStore`. That would make things the easiest.
If we tie the life cycle to the client itself we get two problems:
1. The stats will be off on a per repository basis if the client is reused across repositories
2. If the secret settings are refreshed and the client instances rebuild, then we'd lose the counts.

So as I said in-line: Either we have to find a way of tagging specific requests (probably not possible from what I can see) or we need to key the clients by the client name + repo name and stop sharing them. Then on client reset/reload (see `org.elasticsearch.repositories.gcs.GoogleCloudStorageService#refreshAndClearCache`) we could pass the stats object along for the existing clients for example so that it lives on even if the client is reloaded.

Happy to discuss this on Slack/Zoom if that's helpful for you (the way the client caching works is probably not 100% obvious from just the code).

fcofdez committed:
Fix code style errors.

fcofdez committed:
Address review comments

Use a new Storage instance per clientName -> repository pair instead
of a shared one.

fcofdez commented:
Thanks for the detailed review @original-brownbear. 
I've pushed an additional commit where I address the problem where the same stats are shared across different repositories. I've decided to create a new client per (client, repository) since the interceptor is defined per `Storage` instance and it isn't possible to mark the requests in any way from the high level APIs. 

Additionally, I've decided to move the `GoogleCloudStorageOperationsStats` instances to the blob store and let `org.elasticsearch.repositories.gcs.GoogleCloudStorageService#client` take an additional parameter with those stats. That simplifies the lifecycle management of those stats.

Let me know what you think.

original-brownbear commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -32,25 +36,26 @@
+     * a key created using {@link #clientKey(String, String)}.
      */
-    private final AtomicReference<Map<String, LazyInitializable<Storage, IOException>>> clientsCache = new AtomicReference<>(emptyMap());
+    private volatile Map<String, Storage> clientCache = emptyMap();
```
The approach here looks fine now in general but we need to enhance it to deal with closing repositories somehow don't we?
Otherwise, we get two problems:

1. If we close a repository its client will still be in this map, effectively creating a memory leak? (looks like we're actually making a similar mistake in `org.elasticsearch.repositories.s3.S3Service#derivedClientSettings` that we could fix eventually ... but leaking full `Storage` instances seems a little scarier since they hold on to non-trivial resources with the internal http client and such)
2. If we close a repository and re-create it at a later date, we'll the stats instance from the previous repository right? (see below)

original-brownbear commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -82,41 +77,73 @@ public synchronized void refreshAndClearCache(Map<String, GoogleCloudStorageClie
+                          final GoogleCloudStorageOperationsStats stats) throws IOException {
+        {
+            final Storage storage = clientCache.get(clientKey(clientName, repositoryName));
+            if (storage != null) {
```
See my comment above, if we close and reopen a repository then it looks like we could get an old `Storage` instance here that doesn't use the passed `stats` but rather the stats from the previous repository right?

original-brownbear commented on plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageServiceTests.java:
```diff
@@ -92,31 +97,54 @@ public void testReinitClientSettings() throws Exception {
+            assertThat(repo1Client, is(not(theInstance(repo2Client))));
+            assertThat(repo1Client, is(theInstance(repo1ClientSecondInstance)));
+        }
+
```
NIT: redundant empty line

original-brownbear reviewed (changes requested):
Thanks @fcofdez looks much better now. I think we still need to properly deal with repository closing but other than that, just some smaller comments on details. Thanks!

original-brownbear commented:
@fcofdez BTW, you'll also have to merge latest `master` into this branch to fix the BwC tests I think.

fcofdez committed:
Release GCS storage clients during repository deletion

fcofdez committed:
Merge remote-tracking branch 'origin/master' into gcs-api-calls-tracking

fcofdez commented:
Thanks again for the review @original-brownbear. I totally missed the repository lifecycle, I've fixed that in the last commit.

fcofdez committed:
Clear only the repository client from cache when it is closed.

Make GCS URL trackers more restrictive, including the bucket name
in the URL.

fcofdez committed:
Fix code linter issues.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into gcs-api-calls-tracking

original-brownbear commented on plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageServiceTests.java:
```diff
@@ -92,31 +97,53 @@ public void testReinitClientSettings() throws Exception {
+            final Storage repo1ClientSecondInstance =
+                storageService.client("gcs1", "repo1", new GoogleCloudStorageOperationsStats("bucket"));
+
+            assertThat(repo1Client, is(not(theInstance(repo2Client))));
```
NIT: we have the `assertNotSame(repo1Client, repo2Client)` shortcut here :)

original-brownbear commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -82,41 +75,88 @@ public synchronized void refreshAndClearCache(Map<String, GoogleCloudStorageClie
+        }
+    }
+
+    private String clientKey(String clientName, String repositoryName) {
```
I just realized this working on https://github.com/elastic/elasticsearch/pull/56703, do we even need to do this compound key magic? We can just cache the clients by `repositoryName` and it's all good as far as I can see? We never have two repos by the same name, so we can keep this simple right?

original-brownbear reviewed (approved):
LGTM thanks for the iterations @fcofdez ! Minor open point on the map keys for caching the clients but that doesn't need another round of review.

fcofdez committed:
Use repository name as the key for client cache in GCS.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into gcs-api-calls-tracking
